HW3/P4/median_filter.cl

```diff
@@ -9,26 +19,56 @@ median_3x3(__global __read_only float *in_values,
            int buf_w, int buf_h,
            const int halo)
 {
-    // Note: It may be easier for you to implement median filtering
-    // without using the local buffer, first, then adjust your code to
-    // use such a buffer after you have that working.
+    // Global position of output pixel
+    const int x = get_global_id(0);
+    const int y = get_global_id(1);
 
+    // Local position relative to (0, 0) in workgroup
+    const int lx = get_local_id(0);
+    const int ly = get_local_id(1);
 
-    // Load into buffer (with 1-pixel halo).
-    //
-    // It may be helpful to consult HW3 Problem 5, and
-    // https://github.com/harvard-cs205/OpenCL-examples/blob/master/load_halo.cl
-    //
-    // Note that globally out-of-bounds pixels should be replaced
-    // with the nearest valid pixel's value.
+    
+    // coordinates of the upper left corner of the buffer in image
+    // space, including halo
+    const int buf_corner_x = x - lx - halo;
+    const int buf_corner_y = y - ly - halo;
 
+    // coordinates of our pixel in the local buffer
+    const int buf_x = lx + halo;
+    const int buf_y = ly + halo;
 
     // Compute 3x3 median for each pixel in core (non-halo) pixels
     //
     // We've given you median9.h, and included it above, so you can
     // use the median9() function.
+    int line, row;
+    
+    float middle = 0.0f;
+    
+    for (row = -1; row <= 1; row ++)
+    {
+        for (line = -1; line <= 1; line ++)
+        {
```

Comment:
This code access each memory location multiple times.
You should look again at:
HW3 Problem 5, and https://github.com/harvard-cs205/OpenCL-examples/blob/master/load_halo.cl
to see how it should be done with less global memory accesses.